src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +181,55 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path, bool isNavigationIntercepted)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (OnNavigateAsync == null)
+            {
+                Refresh(isNavigationIntercepted);
+                return;
+            }
+
+            // Create a new CTS for this invocation of the task and
+            // extract the reference to the current one.
+            var pendingOnNavigateCts = _onNavigateCts;
+            var onNavigateCts = new CancellationTokenSource();
+            _onNavigateCts = onNavigateCts;
+
+            var navigateContext = new OnNavigateArgs(path, _onNavigateCts);
+            var task = OnNavigateAsync(navigateContext);
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            if (pendingOnNavigateCts != null)
+            {
+                pendingOnNavigateCts.Cancel();
+            }
+
+            // Create a cancellation task based on the cancellation token
+            // associated with the current running task.
+            var cancellationTaskSource = new TaskCompletionSource();
+            navigateContext.CancellationTokenSource.Token.Register(() =>
+                cancellationTaskSource.TrySetCanceled(navigateContext.CancellationTokenSource.Token));
+
+            // If the user provided a Loading render fragment, then show it.
+            if (Loading != null)
+            {
+                _renderHandle.Render(Loading);
+            }
+
+            await Task.WhenAny(task, cancellationTaskSource.Task);
```

Comment:
It might be possible to eliminate the need for `cancellationTaskSource` by doing this something like:

```cs
await Task.WhenAny(task, Task.Delay(Timeout.Infinite, navigateContext.CancellationTokenSource.Token));
```

Reply:
I would be careful with those things, since I'm not sure how it affects timer registration and other lower level things. I'm not saying that what you are suggesting is wrong, just that I can't confidently say is right.

Reply:
We use the `Task.Delay(infinite, token)` approach in a couple of places in our repo already, though admittedly only tests.

The approach is suggested in [this doc](https://devblogs.microsoft.com/cesardelatorre/implementing-background-tasks-in-microservices-with-ihostedservice-and-the-backgroundservice-class-net-core-2-x/), plus many other places on the internets.

Reply:
I would check with @davidfowl since he's very knowledgeable in this area and I believe has written explicit guidance.

Reply:
Good. I don't know for sure if there are drawbacks with `Timeout.Infinite` so would be good to check.

Reply:
I did try using a `Task.Delay(Timeout.Infinite, token)` to observe the cancellation since I came across the suggestion as well.

It didn't really work, although I was playing around with triggering the cancellation from the callback at that time so there might have been some other things going on there.

Reply:
I'd leave this be. We use the CTS pattern elsewhere in the framework \ blazor, helps to have a consistent strategy for cancelling these.

Reply:
Does this code all run single threaded (no overlapping calls?)

Reply:
Runs single threaded but calls can overlap if they go async

Reply:
I'll try to get some time to look at this but in the meantime @BrennanConroy and @halter73 can you also take a look?

Reply:
> Runs single threaded but calls can overlap if they go async

This seems problematic considering an overlapping call disposes the _onNavigateCts which backs NavigationContext.CancellationToken. If I'm understanding this correctly, that means if the OnNavigateAsync callback tries to register with the CancellationToken after it goes async, it would throw an ODE.

Reply:
Also, since the only `await` is at the very end of the method, the method could be made to not be `async` and just return the Task from WhenAny directly. This saves an allocation of an AsyncStateMachine when this method would otherwise "go async" so-to-speak. It can make stacktraces harder to follow if the task faults though.

`Task.Delay(Timeout.Infinite, navigateContext.CancellationTokenSource.Token)` should not be a problem since it's directly awaited (or returned then awaited). Task.Delay always cleans up after itself disposing any registrations or timers it owns when it completes or is canceled. Task.Delay only becomes problematic when combined with something like Task.WhenAny, because then you might be done with the Task returned by Task.Delay before it ever completes or is canceled.

Reply:
Let's figure this out in a follow up. I'd like for us to get this PR in. @captainsafia could you file an issue to track this? Thanks!

Reply:
The issue for this is here: https://github.com/dotnet/aspnetcore/issues/23789

It seems like the biggest issue to resolve is the way we manage the _onNavigateCts.